esp_custom_part_ota/include/esp_custom_part_ota.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <stdbool.h>
+#include <esp_err.h>
+#include <esp_idf_version.h>
+#include "esp_partition.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef void *esp_custom_part_ota_handle_t;
+
+typedef struct {
+    const esp_partition_t *update_partition;
+    const esp_partition_t *backup_partition;
+} esp_custom_part_ota_cfg_t;
+
+
+/**
+ * @brief   Commence an OTA update writing to the specified partition.
+ *
+ * @param config    Configuration data for the OTA process
+ *
+ * @return
+ *    - NULL    On failure
+ *    - esp_custom_part_ota_handle_t handle
+*/
+esp_custom_part_ota_handle_t esp_custom_part_ota_begin(esp_custom_part_ota_cfg_t config);
+
+
+/**
+ * @brief   Write OTA update data to partition
+ *
+ * This function can be called multiple times as
+ * data is received during the OTA operation. Data is written
+ * sequentially to the partition.
+ *
+ * @param ctx   esp_custom_part_ota_handle_t
+ * @param data  Data buffer to write
+ * @param size  Size of the data buffer in bytes
+ *
+ * @return
+ *      - ESP_OK
+ *      - ESP_ERR_INVALID_ARG
+ *      - lower level esp_partition error codes.
+*/
+esp_err_t esp_custom_part_ota_write(esp_custom_part_ota_handle_t handle, const void *data, size_t size);
+
+
+/**
+ * @brief   Finish OTA update
+ *
+ * @param ctx   esp_custom_part_ota_handle_t
```

Comment:
This API invalidates the backup data if available.
Please note that this does not erase the actual data on the flash. For that you may need to explicitly erase the backup partition